### PR TITLE
Proposal for a new way of configuring webpack

### PR DESCRIPTION
**Pull request by sndrs**
## What does this change?

Following [an RFC](https://docs.google.com/document/d/1iBtzX2cQKSBPxYba5HlxB8NYieGjGvwEFhKh-6RvNOQ/edit#heading=h.ze95s9luaqtz), this proposes configuring webpack so that:

1. there is a one-to-one mapping between webpack bundle configs and built bundles e.g. 
    - `bundle.apps.mjs` -> `dist/apps/*`
    - `bundle.web.mjs` -> `dist/web/*`
    - `bundle.web.variant.mjs` -> `dist/web.variant/*`
1. the main way-in to build the bundles is an orthodox, top level `webpack.config` which composes the bundle configs
3. bundle configs themselves live in the `configs/webpack` directory
4. each bundle config should be a functioning, viable, standalone config
1. building the configs generates a TypeScript manifest of the bundles in the app, for type safety on the server
2. organises `dist` into bundle directories, rather than by filename 

## Why?

- if you know how to use webpack, it should look familiar: [there should be as few surprises/indirections as possible](https://github.com/guardian/recommendations/blob/main/coding-with-empathy.md#recommendations)
- it makes it easier to see how variant bundles differ from the original
- it makes it easier to add/remove variants

## What does this not change?

Everything to do with how we current build bundles. The new configs are not in use and not run by any `make` scripts etc. This just to proposes a solution for now, although they _should_ work.

## What are the differences in the output?

This is the result of running the tests in `configs/webpack/.test.mjs`:

```sh
client.web#name does not match bundle.web#name
│ Expected values to be strictly deep-equal:
│ + actual - expected
│ 
│ + 'client.web'
│ - 'bundle.web'
└───────────
client.web#mode matches bundle.web#mode
client.web#output does not match bundle.web#output
│ Expected values to be strictly deep-equal:
│ + actual - expected
│ 
│   {
│ +   chunkFilename: '[name].web.[chunkhash].js',
│ -   chunkFilename: '[name].[chunkhash].js',
│     filename: [Function: filename],
│ +   path: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist',
│ -   path: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/bundle.web',
│     publicPath: ''
│   }
└───────────
client.web#stats matches bundle.web#stats
client.web#devtool matches bundle.web#devtool
client.web#resolve matches bundle.web#resolve
client.web#ignoreWarnings matches bundle.web#ignoreWarnings
client.web#plugins does not match bundle.web#plugins
│ Expected values to be strictly deep-equal:
│ + actual - expected ... Lines skipped
│ 
│   [
│     DefinePlugin {
│ ...
│         logLevel: 'warn',
│         openAnalyzer: false,
│ +       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/stats/client.web-bundles.html',
│ -       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/bundle.web/stats/bundles.html',
│         reportTitle: [Function (anonymous)],
│         startAnalyzer: true,
│ ...
│         logLevel: 'warn',
│         openAnalyzer: false,
│ +       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/stats/client.web-bundles.json',
│ -       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/bundle.web/stats/bundles.json',
│         reportTitle: [Function (anonymous)],
│         startAnalyzer: true,
│ ...
│         assetHookStage: Infinity,
│         basePath: '',
│ +       fileName: 'manifest.web.json',
│ -       fileName: 'manifest.json',
│         filter: null,
│ ...
│       }
│     }
│   ]
└───────────
client.web#infrastructureLogging matches bundle.web#infrastructureLogging
client.web#entry matches bundle.web#entry
client.web#optimization matches bundle.web#optimization
client.web#module matches bundle.web#module
client.web.scheduled#name does not match bundle.web.scheduled#name
│ Expected values to be strictly deep-equal:
│ + actual - expected
│ 
│ + 'client.web.scheduled'
│ - 'bundle.web.scheduled'
└───────────
client.web.scheduled#mode matches bundle.web.scheduled#mode
client.web.scheduled#output does not match bundle.web.scheduled#output
│ Expected values to be strictly deep-equal:
│ + actual - expected
│ 
│   {
│ +   chunkFilename: '[name].web.scheduled.[chunkhash].js',
│ -   chunkFilename: '[name].[chunkhash].js',
│     filename: [Function: filename],
│ +   path: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist',
│ -   path: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/bundle.web.scheduled',
│     publicPath: ''
│   }
└───────────
client.web.scheduled#stats matches bundle.web.scheduled#stats
client.web.scheduled#devtool matches bundle.web.scheduled#devtool
client.web.scheduled#resolve matches bundle.web.scheduled#resolve
client.web.scheduled#ignoreWarnings matches bundle.web.scheduled#ignoreWarnings
client.web.scheduled#plugins does not match bundle.web.scheduled#plugins
│ Expected values to be strictly deep-equal:
│ + actual - expected ... Lines skipped
│ 
│   [
│     DefinePlugin {
│ ...
│         logLevel: 'warn',
│         openAnalyzer: false,
│ +       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/stats/client.web.scheduled-bundles.html',
│ -       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/bundle.web.scheduled/stats/bundles.html',
│         reportTitle: [Function (anonymous)],
│         startAnalyzer: true,
│ ...
│         logLevel: 'warn',
│         openAnalyzer: false,
│ +       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/stats/client.web.scheduled-bundles.json',
│ -       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/bundle.web.scheduled/stats/bundles.json',
│         reportTitle: [Function (anonymous)],
│         startAnalyzer: true,
│ ...
│         assetHookStage: Infinity,
│         basePath: '',
│ +       fileName: 'manifest.web.scheduled.json',
│ -       fileName: 'manifest.json',
│         filter: null,
│ ...
│       }
│     }
│   ]
└───────────
client.web.scheduled#infrastructureLogging matches bundle.web.scheduled#infrastructureLogging
client.web.scheduled#entry matches bundle.web.scheduled#entry
client.web.scheduled#optimization matches bundle.web.scheduled#optimization
client.web.scheduled#module matches bundle.web.scheduled#module
client.web.variant#name does not match bundle.web.variant#name
│ Expected values to be strictly deep-equal:
│ + actual - expected
│ 
│ + 'client.web.variant'
│ - 'bundle.web.variant'
└───────────
client.web.variant#mode matches bundle.web.variant#mode
client.web.variant#output does not match bundle.web.variant#output
│ Expected values to be strictly deep-equal:
│ + actual - expected
│ 
│   {
│ +   chunkFilename: '[name].web.variant.[chunkhash].js',
│ -   chunkFilename: '[name].[chunkhash].js',
│     filename: [Function: filename],
│ +   path: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist',
│ -   path: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/bundle.web.variant',
│     publicPath: ''
│   }
└───────────
client.web.variant#stats matches bundle.web.variant#stats
client.web.variant#devtool matches bundle.web.variant#devtool
client.web.variant#resolve matches bundle.web.variant#resolve
client.web.variant#ignoreWarnings matches bundle.web.variant#ignoreWarnings
client.web.variant#plugins does not match bundle.web.variant#plugins
│ Expected values to be strictly deep-equal:
│ + actual - expected ... Lines skipped
│ 
│   [
│     DefinePlugin {
│ ...
│         logLevel: 'warn',
│         openAnalyzer: false,
│ +       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/stats/client.web.variant-bundles.html',
│ -       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/bundle.web.variant/stats/bundles.html',
│         reportTitle: [Function (anonymous)],
│         startAnalyzer: true,
│ ...
│         logLevel: 'warn',
│         openAnalyzer: false,
│ +       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/stats/client.web.variant-bundles.json',
│ -       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/bundle.web.variant/stats/bundles.json',
│         reportTitle: [Function (anonymous)],
│         startAnalyzer: true,
│ ...
│         assetHookStage: Infinity,
│         basePath: '',
│ +       fileName: 'manifest.web.variant.json',
│ -       fileName: 'manifest.json',
│         filter: null,
│ ...
│       }
│     }
│   ]
└───────────
client.web.variant#infrastructureLogging matches bundle.web.variant#infrastructureLogging
client.web.variant#entry matches bundle.web.variant#entry
client.web.variant#optimization matches bundle.web.variant#optimization
client.web.variant#module matches bundle.web.variant#module
client.web.legacy#name does not match bundle.web.legacy#name
│ Expected values to be strictly deep-equal:
│ + actual - expected
│ 
│ + 'client.web.legacy'
│ - 'bundle.web.legacy'
└───────────
client.web.legacy#mode matches bundle.web.legacy#mode
client.web.legacy#output does not match bundle.web.legacy#output
│ Expected values to be strictly deep-equal:
│ + actual - expected
│ 
│   {
│ +   chunkFilename: '[name].web.legacy.[chunkhash].js',
│ -   chunkFilename: '[name].[chunkhash].js',
│     filename: [Function: filename],
│ +   path: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist',
│ -   path: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/bundle.web.legacy',
│     publicPath: ''
│   }
└───────────
client.web.legacy#stats matches bundle.web.legacy#stats
client.web.legacy#devtool matches bundle.web.legacy#devtool
client.web.legacy#resolve matches bundle.web.legacy#resolve
client.web.legacy#ignoreWarnings matches bundle.web.legacy#ignoreWarnings
client.web.legacy#plugins does not match bundle.web.legacy#plugins
│ Expected values to be strictly deep-equal:
│ + actual - expected ... Lines skipped
│ 
│   [
│     DefinePlugin {
│ ...
│         logLevel: 'warn',
│         openAnalyzer: false,
│ +       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/stats/client.web.legacy-bundles.html',
│ -       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/bundle.web.legacy/stats/bundles.html',
│         reportTitle: [Function (anonymous)],
│         startAnalyzer: true,
│ ...
│         logLevel: 'warn',
│         openAnalyzer: false,
│ +       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/stats/client.web.legacy-bundles.json',
│ -       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/bundle.web.legacy/stats/bundles.json',
│         reportTitle: [Function (anonymous)],
│         startAnalyzer: true,
│ ...
│         assetHookStage: Infinity,
│         basePath: '',
│ +       fileName: 'manifest.web.legacy.json',
│ -       fileName: 'manifest.json',
│         filter: null,
│ ...
│       }
│     }
│   ]
└───────────
client.web.legacy#infrastructureLogging matches bundle.web.legacy#infrastructureLogging
client.web.legacy#entry matches bundle.web.legacy#entry
client.web.legacy#optimization matches bundle.web.legacy#optimization
client.web.legacy#module matches bundle.web.legacy#module
client.apps#name does not match bundle.apps#name
│ Expected values to be strictly deep-equal:
│ + actual - expected
│ 
│ + 'client.apps'
│ - 'bundle.apps'
└───────────
client.apps#mode matches bundle.apps#mode
client.apps#output does not match bundle.apps#output
│ Expected values to be strictly deep-equal:
│ + actual - expected
│ 
│   {
│ +   chunkFilename: '[name].apps.[chunkhash].js',
│ -   chunkFilename: '[name].[chunkhash].js',
│     filename: [Function: filename],
│ +   path: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist',
│ -   path: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/bundle.apps',
│     publicPath: ''
│   }
└───────────
client.apps#stats matches bundle.apps#stats
client.apps#devtool matches bundle.apps#devtool
client.apps#resolve matches bundle.apps#resolve
client.apps#ignoreWarnings matches bundle.apps#ignoreWarnings
client.apps#plugins does not match bundle.apps#plugins
│ Expected values to be strictly deep-equal:
│ + actual - expected ... Lines skipped
│ 
│   [
│     DefinePlugin {
│ ...
│         logLevel: 'warn',
│         openAnalyzer: false,
│ +       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/stats/client.apps-bundles.html',
│ -       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/bundle.apps/stats/bundles.html',
│         reportTitle: [Function (anonymous)],
│         startAnalyzer: true,
│ ...
│         logLevel: 'warn',
│         openAnalyzer: false,
│ +       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/stats/client.apps-bundles.json',
│ -       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/bundle.apps/stats/bundles.json',
│         reportTitle: [Function (anonymous)],
│         startAnalyzer: true,
│ ...
│         assetHookStage: Infinity,
│         basePath: '',
│ +       fileName: 'manifest.apps.json',
│ -       fileName: 'manifest.json',
│         filter: null,
│ ...
│       }
│     }
│   ]
└───────────
client.apps#infrastructureLogging matches bundle.apps#infrastructureLogging
client.apps#entry matches bundle.apps#entry
client.apps#optimization matches bundle.apps#optimization
client.apps#module matches bundle.apps#module
server#name does not match bundle.server#name
│ Expected values to be strictly deep-equal:
│ + actual - expected
│ 
│ + 'server'
│ - 'bundle.server'
└───────────
server#mode matches bundle.server#mode
server#output does not match bundle.server#output
│ Expected values to be strictly deep-equal:
│ + actual - expected
│ 
│   {
│     chunkFilename: '[name].js',
│     filename: '[name].js',
│     libraryTarget: 'commonjs2',
│ +   path: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist',
│ -   path: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/bundle.server',
│     pathinfo: true,
│ -   publicPath: ''
│   }
└───────────
server#stats matches bundle.server#stats
server#devtool matches bundle.server#devtool
server#resolve matches bundle.server#resolve
server#ignoreWarnings matches bundle.server#ignoreWarnings
server#plugins does not match bundle.server#plugins
│ Expected values to be strictly deep-equal:
│ + actual - expected ... Lines skipped
│ 
│   [
│     DefinePlugin {
│ ...
│         logLevel: 'warn',
│         openAnalyzer: false,
│ +       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/stats/server-bundles.html',
│ -       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/bundle.server/stats/bundles.html',
│         reportTitle: [Function (anonymous)],
│         startAnalyzer: true,
│ ...
│         logLevel: 'warn',
│         openAnalyzer: false,
│ +       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/stats/server-bundles.json',
│ -       reportFilename: '/Users/alex_sanders/Documents/code/dotcom-rendering/dotcom-rendering/dist/bundle.server/stats/bundles.json',
│         reportTitle: [Function (anonymous)],
│ ...
│       server: null
│     }
│   ]
└───────────
server#infrastructureLogging matches bundle.server#infrastructureLogging
server#entry matches bundle.server#entry
server#target matches bundle.server#target
server#externalsPresets matches bundle.server#externalsPresets
server#optimization matches bundle.server#optimization
server#externals does not match bundle.server#externals
│ Expected values to be strictly deep-equal:
│ + actual - expected
│ 
│   [
│ +   [Function (anonymous)],
│ +   [Function (anonymous)],
│     [Function (anonymous)]
│   ]
└───────────
server#module matches bundle.server#module
```

## Known issues

The reorganising of `dist` has some issues with the dev server (#8767). This will need some looking into.

_Co-authored with @mxdvl._

### CODE REVIEW

**Review by arelra — APPROVED** [review]
This is a valuable proposal which also sets us up for further splitting up of the application. Thanks for raising it. 

It would be a good to get it finalised. I see the following points mentioned as remaining work?:

- the dev server
- assets are not served correctly

File: dotcom-rendering/webpack.config.mjs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+const bundleConfigs = await readdir(configsDirectory).then((files) =>
+	files.filter((file) => file.startsWith('bundle')),
+);
+
+// Create a bundles manifest for the TypesScript compiler so we get type safely
+// when referring to bundles on the server
+await mkdir(resolve(dirname, 'dist'), { recursive: true });
+await writeFile(
+	resolve(dirname, 'dist', 'manifest.bundles.js'),
+	`// DO NOT EDIT: THIS FILE IS AUTOMATICALLY GENERATED //
```
[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/webpack.config.mjs:33]: Yes, but that specific way will not work in modules 😉!

File: dotcom-rendering/configs/webpack/bundle.server.dev.mjs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+				: callback();
+		},
+	],
+	module: {
+		rules: [
+			{
+				test: /(\.tsx|\.js|\.ts)$/,
+				use: [
+					{
+						loader: 'swc-loader',
```
[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/configs/webpack/bundle.server.dev.mjs:49]: In this case, this is the only “change” applied to the `server` config. In practice, I think that the DEV server should have as few modifications from the PROD server, and the shortness of this file really expresses that.

Generating the exact configuration should be relatively trivial: it’s only a matter of importing the default export and displaying it.

File: dotcom-rendering/webpack.config.mjs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+const dirname = fileURLToPath(new URL('.', import.meta.url));
+
+const configsDirectory = resolve(dirname, 'configs', 'webpack');
+
+const bundleConfigs = await readdir(configsDirectory).then((files) =>
+	files.filter((file) => file.startsWith('bundle')),
+);
+
+// Create a bundles manifest for the TypesScript compiler so we get type safely
+// when referring to bundles on the server
```
[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/webpack.config.mjs:29]: Yes, in [`src/lib/assets.ts`](https://github.com/guardian/dotcom-rendering/blob/9bece8c5311b01eb60644b559bc677c62ec4c880/dotcom-rendering/src/lib/assets.ts#L66) we want to ensure that only valid “builds” are considered.

File: dotcom-rendering/configs/webpack/bundle.web.scheduled.mjs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+/** @type {import("webpack").Configuration} */
+export const webScheduled = {
+	entry: {
+		index: './src/client/index.scheduled.ts',
+		debug: './src/client/debug/index.ts',
+	},
+};
+
+export default merge(base(import.meta.url), web, webScheduled);
```
[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/configs/webpack/bundle.web.scheduled.mjs:22]: That can be done, but is unneccessary for the purpose of this PR. We thought the test were sufficient to confirm that the output matches the current configuration.

File: dotcom-rendering/webpack.config.mjs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import { mkdir, readdir, writeFile } from 'node:fs/promises';
+import { basename, resolve } from 'node:path';
+import { fileURLToPath } from 'node:url';
+
+const dirname = fileURLToPath(new URL('.', import.meta.url));
+
+const configsDirectory = resolve(dirname, 'configs', 'webpack');
+
+const bundleConfigs = await readdir(configsDirectory).then((files) =>
+	files.filter((file) => file.startsWith('bundle')),
```
[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/webpack.config.mjs:25]: No strong feelings here, but we thought this little redundancy came with the benefit of being very explicit. These files define a “bundle”, and the fact that they are  Webpack configurations is more of an implementation detail.

File: dotcom-rendering/configs/webpack/bundle.server.dev.mjs (hunk 2 above)
[the remark this answers is not in the document]
Comment (arelra) [reply dotcom-rendering/configs/webpack/bundle.server.dev.mjs:49]: I think splitting the configuration entries  across the server and dev server means it is less explicit and harder to determine for the reader. Aiui this is a scenario we wanted to move away from.

Would an option be to not split/merge properties like the swc configuration?

Not a blocker.

File: dotcom-rendering/configs/webpack/bundle.web.scheduled.mjs (hunk 4 above)
[the remark this answers is not in the document]
Comment (arelra) [reply dotcom-rendering/configs/webpack/bundle.web.scheduled.mjs:22]: Nice to have not a blocker

**Pull request merged**